README.md
# Frog Auth

Authentication tools for [Dart Frog](http://dartfrog.vgv.dev) apps.

## Installation

![Pub Version](https://img.shields.io/pub/v/frog_auth)

```sh
dart pub add frog_auth
```

## Usage

### Basic Authentication

To support [Basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication), simply add the `basicAuthentication` middleware:

```dart
Handler middleware(Handler handler) {
    return handler.use(
        basicAuthentication(
            retrieveUser: (context, username, password) async {
                // TODO Retrieve user by username/password
            },
        ),
    );
}
```

The `retrieveUser` callback should be used to lookup the user using the given username and password. If no user is found with the given credentials, you should return `null`.

If a non-null user is returned by the `retrieveUser` callback, it will be provided to the current request context and can be retrieved using `context.read()`.

`retrieveUser` can return an object of any type extending `Object`, so should be flexible enough to work with any database system.

### Bearer Authentication

To support [Bearer authentication](https://datatracker.ietf.org/doc/html/rfc6750), simply add the `bearerAuthentication` middleware:

```dart
Handler middleware(Handler handler) {
    return handler.use(
        bearerAuthentication(
            retrieveUser: (context, token) async {
                // TODO Retrieve user by token
            },
        ),
    );
}
```

The `retrieveUser` callback should be used to lookup the user using the given token. If no user is found with the given token, you should return `null`.

If a non-null user is returned by the `retrieveUser` callback, it will be provided to the current request context and can be retrieved using `context.read()`.

`retrieveUser` can return an object of any type extending `Object`, so should be flexible enough to work with any database system.


CHANGELOG.md
## 0.0.3

- Updated to dart_frog 0.0.2

## 0.0.2

- Provide `RequestContext` in `retrieveUser` callbacks

## 0.0.1

- Support HTTP Basic authentication
- Support Bearer token authentication

test/frog_auth_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:dart_frog/dart_frog.dart';
import 'package:frog_auth/frog_auth.dart';
import 'package:mocktail/mocktail.dart';
import 'package:test/test.dart';

class _MockRequestContext extends Mock implements RequestContext {}

void main() {
  Response onRequest(RequestContext _) {
    return Response(body: 'Hello, World!');
  }

  late _MockRequestContext context;

  setUp(() {
    context = _MockRequestContext();
    when(() => context.provide<Map<String, dynamic>>(any()))
        .thenReturn(context);
  });

  group('basic authentication', () {
    late Handler handler;

    setUp(() {
      handler = const Pipeline().addMiddleware(
        basicAuthentication(
          retrieveUser: (_, username, password) async {
            return username == 'peter' && password == 'test'
                ? <String, dynamic>{'username': username}
                : null;
          },
        ),
      ).addHandler(onRequest);
    });

    test('returns ok when user is found', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'Basic ${base64Encode(utf8.encode('peter:test'))}',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.ok));
      await expectLater(await response.body(), equals('Hello, World!'));
    });

    test('returns 401 when user credentials are invalid', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'Basic ${base64Encode(utf8.encode('peter:wrong'))}',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.unauthorized));
    });

    test('returns 400 when credentials do not start with "Basic"', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'Bearer ${base64Encode(utf8.encode('peter:test'))}',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.badRequest));
    });

    test('returns 400 when Authorization header is malformed', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'peter:test',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.badRequest));
    });

    test('returns 400 when Authorization header is missing', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {},
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.badRequest));
    });
  });

  group('bearer authentication', () {
    late Handler handler;

    setUp(() {
      handler = const Pipeline().addMiddleware(
        bearerAuthentication(
          retrieveUser: (_, token) async {
            return token == 'test_token'
                ? <String, dynamic>{'username': 'test'}
                : null;
          },
        ),
      ).addHandler(onRequest);
    });

    test('returns ok when valid token is provided', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'Bearer test_token',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.ok));
      await expectLater(await response.body(), equals('Hello, World!'));
    });

    test('returns 401 when token is invalid', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'Bearer another_token',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.unauthorized));
    });

    test(
      'returns 400 when Authorization header does not start with "Bearer"',
      () async {
        final request = Request.get(
          Uri.parse('http://localhost/'),
          headers: {
            'Authorization': 'Basic peter:test',
          },
        );

        when(() => context.request).thenReturn(request);

        final response = await handler(context);

        await expectLater(response.statusCode, equals(HttpStatus.badRequest));
      },
    );

    test('returns 400 when Authorization header is malformed', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {
          'Authorization': 'peter:test',
        },
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.badRequest));
    });

    test('returns 400 when Authorization header is missing', () async {
      final request = Request.get(
        Uri.parse('http://localhost/'),
        headers: {},
      );

      when(() => context.request).thenReturn(request);

      final response = await handler(context);

      await expectLater(response.statusCode, equals(HttpStatus.badRequest));
    });
  });
}


lib/src/bearer_auth.dart
import 'dart:io';

import 'package:dart_frog/dart_frog.dart';

/// A Bearer authentication scheme.
///
/// This is a simple authentication scheme that uses a token provided in the
/// Authorization header. The header should be in the form "Bearer <token>".
///
/// Requires a [retrieveUser] callback, which you should use to look up the
/// user in your database. The callback should return `null` if no user is
/// found; in that case the middleware will response with a 401 (Unauthorized).
///
/// If the user is found, the middleware will continue the request and provide
/// the retrieved user object to the request context. It can be accessed using
/// `context.read<User>()`.
///
/// The middleware will also respond with a 400 (Bad Request) if the
/// Authorization header is missing or malformed.
Middleware bearerAuthentication<User extends Object>({
  required UserFromToken<User> retrieveUser,
}) {
  return (handler) {
    return (context) async {
      final authHeader = context.request.headers['Authorization'];

      if (authHeader == null || !authHeader.startsWith('Bearer ')) {
        return Response(statusCode: HttpStatus.badRequest);
      }

      try {
        final token = authHeader.split(' ')[1];
        final user = await retrieveUser(context, token);

        if (user == null) {
          return Response(statusCode: HttpStatus.unauthorized);
        }

        return await handler(context.provide(() => user));
      } on Exception {
        return Response(statusCode: HttpStatus.badRequest);
      }
    };
  };
}

typedef UserFromToken<User extends Object> = Future<User?> Function(
  RequestContext context,
  String token,
);


lib/src/basic_auth.dart
import 'dart:convert';
import 'dart:io';

import 'package:dart_frog/dart_frog.dart';

/// A Basic authentication scheme.
///
/// This is a very simple authentication scheme that uses a username and
/// password provided in the Authorization header. The header should be
/// in the form "Basic <username>:<password>", where the username/password
/// portion is base64-encoded.
///
/// Requires a [retrieveUser] callback, which you should use to look up the
/// user in your database. The callback should return `null` if no user is
/// found; in that case the middleware will response with a 401 (Unauthorized).
///
/// If the user is found, the middleware will continue the request and provide
/// the retrieved user object to the request context. It can be accessed using
/// `context.read<User>()`.
///
/// The middleware will also respond with a 400 (Bad Request) if the
/// Authorization header is missing or malformed.
Middleware basicAuthentication<User extends Object>({
  required UserFromCredentials<User> retrieveUser,
}) {
  return (handler) {
    return (context) async {
      final authHeader = context.request.headers['Authorization'];

      if (authHeader == null || !authHeader.startsWith('Basic ')) {
        return Response(statusCode: HttpStatus.badRequest);
      }

      try {
        final credentials = utf8
            .decode(
              base64.decode(authHeader.split(' ')[1]),
            )
            .split(':');

        final username = credentials.first;
        final password = credentials[1];

        final user = await retrieveUser(context, username, password);

        if (user == null) {
          return Response(statusCode: HttpStatus.unauthorized);
        }

        return await handler(context.provide(() => user));
      } on Exception {
        return Response(statusCode: HttpStatus.badRequest);
      }
    };
  };
}

typedef UserFromCredentials<User extends Object> = Future<User?> Function(
  RequestContext context,
  String username,
  String password,
);
